README.md
# 🌌 ⌨️ HyperKey.spoon

If you have a lot of global key binds to `super`, `hyper`, or other modifier
keys, this [Hammerspoon](https://www.hammerspoon.org/docs/index.html) library lets you easily setup your key binds to switch/launch
applications or call a function.

As well, if you hold down the modifier key(s), a popup help menu will appear
showing you all your key binds:

<img src="https://github.com/dbalatero/HyperKey.spoon/raw/master/images/screenshot.png" />

This was inspired by the Spacemacs [spacebar leader menu](https://www.spacemacs.org/doc/QUICK_START.html#the-leader-keys), and [vim-leader-guide](https://github.com/hecal3/vim-leader-guide).

## Installation

The easiest thing to do is paste this in:

```
mkdir -p ~/.hammerspoon/Spoons
git clone https://github.com/dbalatero/HyperKey.spoon.git ~/.hammerspoon/Spoons/HyperKey.spoon
```

then move onto the next section for configuration.

## Usage

```lua
local hyper = {'cmd', 'alt', 'ctrl', 'shift'}

-- Load and create a new switcher
local HyperKey = hs.loadSpoon("HyperKey")
hyperKey = HyperKey:new(hyper)

-- Bind some applications to keys
hyperKey
  :bind('c'):toApplication('/Applications/Google Chrome.app')
  :bind('s'):toApplication('/Applications/Spotify.app')
  :bind('t'):toApplication('/Applications/Alacritty.app')

-- Bind some functions to keys
local reloadHammerspoon = function()
  hs.application.launchOrFocus("Hammerspoon")
  hs.reload()
end

hyperKey
  :bind('h'):toFunction("Reload Hammerspoon", reloadHammerspoon)
  :bind('l'):toFunction("Lock screen", hs.caffeinate.startScreensaver)
```

### Adding a hyper key to OS X

If you don't have a QMK-based keyboard, you can easily add a `hyper` key to
MacOS using Karabiner Elements. There are many tutorials out there; [here's
one](https://brettterpstra.com/2017/06/15/a-hyper-key-with-karabiner-elements-full-instructions/#:~:text=Karabiner%20Elements%20should%20immediately%20detect,%E2%8C%98%E2%87%A7%E2%8C%A5%E2%8C%83X%20)
that might work.

### Controlling the popup delay

The popup delay defaults to `250ms`. You can control this with an option:

```lua
local hyper = {'cmd', 'alt', 'ctrl', 'shift'}

hyperKey = HyperKey:new(hyper, {
  overlayTimeoutMs = 1000, -- wait 1000ms instead
})
```

### Multiple popup menus

Having different menus for different modifiers is easy - just create an object for each set of modifier keys:

```lua
hyperKey = HyperKey:new({'cmd', 'alt', 'ctrl', 'shift'})
hyperKey:bind('s'):toApplication('/Applications/Safari.app')

superKey = HyperKey:new({'cmd', 'alt', 'ctrl'})
superKey:bind('L'):toFunction('Lock screen', hs.caffeinate.startScreensaver)
```

## Methods

### `HyperKey:new(modifiers, options)`

Creates a new set of hotkeys triggered by `modifiers`. Holding the `modifiers`
down will show a popup overlay of all registered keybinds.

* `modifiers` - a Lua table of modifiers, e.g. `{'cmd','shift','alt','ctrl'}`
* `options` - a Lua table of options, with the following keys:
  * `overlayTimeoutMs: number` - number of milliseconds to wait before fading in the hotkey overlay. Defaults to `250`ms.

Usage:

```lua
hyperKey = HyperKey:new(
  {'cmd', 'shift', 'alt', 'ctrl'},
  { overlayTimeoutMs = 1000 }
)

-- then bind some keys to it
```

### `HyperKey:bind(displayedKey, [bindKey]):toFunction(name, fn)`

Binds `modifiers` + `bindKey` to a given `fn` to be called.

Returns `self`, so you can chain keybinds.

* `displayedKey` - the key character to display on the popup.
* `bindKey` - the Hammerspoon key to bind. Defaults to `displayedKey`.
  * This is the same set of key values that `hs.hotkey.bind` takes.
* `name` - the description of this key bind to display on the popup.
* `fn` - the function to call when the hotkey is pressed.

Usage:

```lua
local function doSomethingCool()
  -- fill me in
end

hyperKey = HyperKey:new({'cmd', 'shift'})

hyperKey
  :bind('c'):toFunction("Do a cool thing", doSomethingCool)
```

### `HyperKey:bind(displayedKey, [bindKey]):toApplication(applicationPath)`

Binds `modifiers` + `bindKey` to quick-switch to a given application. If the application is not running, it will launch it.

Returns `self`, so you can chain keybinds.

* `displayedKey` - the key character to display on the popup.
* `bindKey` - the Hammerspoon key to bind. Defaults to `displayedKey`.
  * This is the same set of key values that `hs.hotkey.bind` takes.
* `applicationPath` - the application to launch/switch, e.g. `/Applications/Safari.app`

Usage:

```lua
hyperKey = HyperKey:new({'cmd', 'shift'})

hyperKey
  :bind('m'):toApplication('/Applications/Mail.app')
  :bind('s'):toApplication('/Applications/Safari.app')
```


CHANGELOG.md
# Next

- [bugfix] Redraw the canvas every time, to prevent the garbage collector from eventually deleting our hs.canvas object.

# 1.0.0

- Initial release!


init.lua
-- Setup load paths.
local function scriptPath()
  local str = debug.getinfo(2, "S").source:sub(2)
  return str:match("(.*/)")
end

local rootDir = scriptPath()
package.path = rootDir .. "lib/?.lua;" .. package.path

-----------------------------------------

local newclass = require('HyperKey.utils.yaci')
local onModifierHold = require('HyperKey.utils.on_modifier_hold')
local bindings = require('HyperKey.bindings')
local Overlay = require('HyperKey.overlay')

----------------------------------------------------------------

local HyperKey = newclass("HyperKey")

HyperKey.author = "David Balatero <d@balatero.com>"
HyperKey.homepage = "https://github.com/dbalatero/HyperKey.spoon"
HyperKey.license = "MIT"
HyperKey.name = "HyperKey"
HyperKey.version = "1.0.0"
HyperKey.spoonPath = rootDir

function HyperKey:init(hyperMods, options)
  options = options or {}

  self.overlayTimeoutMs = options.overlayTimeoutMs or 250
  self.hyperMods = hyperMods
  self.bindings = {}
  self.overlay = Overlay:new(self.bindings)
  self.holdTap = self:_createOverlayTap()
end

function HyperKey:bind(displayedKey, bindKey)
  bindKey = bindKey or displayedKey

  return {
    toApplication = function(_, applicationName)
      return self:_bind(
        displayedKey,
        bindKey,
        bindings.ApplicationBinding:new(applicationName)
      )
    end,
    toFunction = function(_, name, fn)
      return self:_bind(
        displayedKey,
        bindKey,
        bindings.FunctionBinding:new(name, fn)
      )
    end
  }
end

function HyperKey:_bind(key, bindKey, binding)
  table.insert(self.bindings, {
    key = string.upper(key),
    bindKey = bindKey,
    binding = binding
  })

  self.overlay = Overlay:new(self.bindings)

  hs.hotkey.bind(self.hyperMods, bindKey, function()
    binding:launch()
  end)

  return self
end

function HyperKey:_createOverlayTap()
  local onHold = function()
    self.overlay:show()
  end

  local onRelease = function()
    self.overlay:hide()
  end

  return onModifierHold(self.hyperMods, self.overlayTimeoutMs, onHold, onRelease)
end

return HyperKey


lib/HyperKey/bindings.lua
local newclass = require('HyperKey.utils.yaci')

local KeyBinding = newclass("KeyBinding")

function KeyBinding:init(name)
  self.name = name
end

function KeyBinding:launch()
  error "Subclass this method instead."
end

function KeyBinding:getImage()
  return nil
end

------

local ApplicationBinding = KeyBinding:subclass("ApplicationBinding")

function ApplicationBinding:init(applicationPath)
  local parts = hs.fnutils.split(applicationPath, "/")
  local name = parts[#parts]

  local nameParts = hs.fnutils.split(name, ".", nil, true)
  local basename = nameParts[1]

  self.super:init(basename)
  self.applicationPath = applicationPath
end

function ApplicationBinding:launch()
  hs.application.launchOrFocus(self.applicationPath)
end

function ApplicationBinding:getImage()
  local bundleInfo = hs.application.infoForBundlePath(self.applicationPath)
  if not bundleInfo then return nil end

  local identifier = bundleInfo.CFBundleIdentifier
  if not identifier then return nil end

  return hs.image.imageFromAppBundle(identifier)
end

------

local FunctionBinding = KeyBinding:subclass('FunctionBinding')

function FunctionBinding:init(name, fn)
  self.super:init(name)
  self.fn = fn
end

function FunctionBinding:launch()
  self.fn()
end

------

return {
  ApplicationBinding = ApplicationBinding,
  FunctionBinding = FunctionBinding,
}


lib/HyperKey/utils/yaci.lua
-----------------------------------------------------------------------------------
-- Yet Another Class Implementation (version 1.2)
--
-- Julien Patte [julien.patte AT gmail DOT com] - 25 Feb 2007
--
-- Inspired from code written by Kevin Baca, Sam Lie, Christian Lindig and others
-- Thanks to Damian Stewart and Frederic Thomas for their interest and comments
-----------------------------------------------------------------------------------

do	-- keep local things inside

-- associations between an object an its meta-informations
-- e.g its class, its "lower" object (if any), ...
local metaObj = {}
setmetatable(metaObj, {__mode = "k"})

-----------------------------------------------------------------------------------
-- internal function 'duplicate'
-- return a shallow copy of table t

local function duplicate(t)
  t2 = {}
  for k,v in pairs(t) do t2[k] = v end
  return t2
end

-----------------------------------------------------------------------------------
-- internal function 'newInstance'

local function newInstance(class, ...)

  local function makeInstance(class, virtuals)
    local inst = duplicate(virtuals)
    metaObj[inst] = { obj = inst, class = class }

    if class:super()~=nil then
      inst.super = makeInstance(class:super(), virtuals)
      metaObj[inst].super = metaObj[inst.super]	-- meta-info about inst
      metaObj[inst.super].lower = metaObj[inst]
    else
      inst.super = {}
    end

    setmetatable(inst, class.static)

    return inst
  end

  local inst = makeInstance(class, metaObj[class].virtuals)
  inst:init(...)
  return inst
end

-----------------------------------------------------------------------------------
-- internal function 'makeVirtual'

local function makeVirtual(class, fname)
  local func = class.static[fname]
  if func == nil then
    func = function() error("Attempt to call an undefined abstract method '"..fname.."'") end
   end
  metaObj[class].virtuals[fname] = func
end

-----------------------------------------------------------------------------------
-- internal function 'trycast'
-- try to cast an instance into an instance of one of its super- or subclasses

local function tryCast(class, inst)
  local meta = metaObj[inst]
  if meta.class==class then return inst end -- is it already the right class?

  while meta~=nil do	-- search lower in the hierarchy
    if meta.class==class then return meta.obj end
    meta = meta.lower
  end

  meta = metaObj[inst].super  -- not found, search through the superclasses
  while meta~=nil do
    if meta.class==class then return meta.obj end
    meta = meta.super
  end

  return nil -- could not execute casting
end

-----------------------------------------------------------------------------------
-- internal function 'secureCast'
-- same as trycast but raise an error in case of failure

local function secureCast(class, inst)
  local casted = tryCast(class, inst)
  if casted == nil then
	error("Failed to cast " .. tostring(inst) .. " to a " .. class:name())
  end
  return casted
end

-----------------------------------------------------------------------------------
-- internal function 'classMade'

local function classMade(class, obj)
  if metaObj[obj]==nil then return false end -- is this really an object?
  return (tryCast(class,obj) ~= nil) -- check if that class could cast the object
end


-----------------------------------------------------------------------------------
-- internal function 'callup'
-- Function used to transfer a method call from a class to its superclass

local callup_inst
local callup_target

local function callup(inst, ...)
  return callup_target(callup_inst, ...)	-- call the superclass' method
end


-----------------------------------------------------------------------------------
-- internal function 'subclass'

local function inst_init_def(inst,...)
  inst.super:init()
end

local function inst_newindex(inst,key,value)
  if inst.super[key] ~= nil then 	-- First check if this field isn't already
									-- defined higher in the hierarchy
	inst.super[key] = value;		-- Update the old value
  else
  	rawset(inst,key,value); 		-- Create the field
  end
end

local function subclass(baseClass, name)
  if type(name)~="string" then name = "Unnamed" end

  local theClass = {}

	-- need to copy everything here because events can't be found through metatables
  local b = baseClass.static
  local inst_stuff = { __tostring=b.__tostring, __eq=b.__eq, __add=b.__add, __sub=b.__sub,
	__mul=b.__mul, __div=b.__div, __mod=b.__mod, __pow=b.__pow, __unm=b.__unm,
	__len=b.__len, __lt=b.__lt, __le=b.__le, __concat=b.__concat, __call=b.__call}

  inst_stuff.init = inst_init_def
  inst_stuff.__newindex = inst_newindex
  function inst_stuff.class() return theClass end

  function inst_stuff.__index(inst, key) -- Look for field 'key' in instance 'inst'
	local res = inst_stuff[key] 		-- Is it present?
	if res~=nil then return res end		-- Okay, return it

	res = inst.super[key]  				-- Is it somewhere higher in the hierarchy?

	if type(res)=='function' and
		res ~= callup then 				-- If it is a method of the superclass,
		callup_inst = inst.super  		-- we will need to do a special forwarding
		callup_target = res  			-- to call 'res' with the correct 'self'
		return callup 					-- The 'callup' function will do that
	end

	return res
  end


  local class_stuff = { static = inst_stuff, made = classMade, new = newInstance,
	subclass = subclass, virtual = makeVirtual, cast = secureCast, trycast = tryCast }
  metaObj[theClass] = { virtuals = duplicate(metaObj[baseClass].virtuals) }

  function class_stuff.name(class) return name end
  function class_stuff.super(class) return baseClass end
  function class_stuff.inherits(class, other)
	return (baseClass==other or baseClass:inherits(other))
  end

  local function newmethod(class, name, meth)
	inst_stuff[name] = meth;
	if metaObj[class].virtuals[name]~=nil then
		metaObj[class].virtuals[name] = meth
	end
  end

  local function tos() return ("class "..name) end
  setmetatable(theClass, { __newindex = newmethod, __index = class_stuff,
	__tostring = tos, __call = newInstance } )

  return theClass
end

-----------------------------------------------------------------------------------
-- The 'Object' class

local Object = {}

local function obj_newitem() error "May not modify the class 'Object'. Subclass it instead." end
local obj_inst_stuff = {}
function obj_inst_stuff.init(inst,...) end
obj_inst_stuff.__index = obj_inst_stuff
obj_inst_stuff.__newindex = obj_newitem
function obj_inst_stuff.class() return Object end
function obj_inst_stuff.__tostring(inst) return ("a "..inst:class():name()) end

local obj_class_stuff = { static = obj_inst_stuff, made = classMade, new = newInstance,
	subclass = subclass, cast = secureCast, trycast = tryCast }
function obj_class_stuff.name(class) return "Object" end
function obj_class_stuff.super(class) return nil end
function obj_class_stuff.inherits(class, other) return false end
metaObj[Object] = { virtuals={} }

local function tos() return ("class Object") end
setmetatable(Object, { __newindex = obj_newitem, __index = obj_class_stuff,
	__tostring = tos, __call = newInstance } )

----------------------------------------------------------------------
-- function 'newclass'

local function newclass(name, baseClass)
 baseClass = baseClass or Object
 return baseClass:subclass(name)
end

return newclass
end -- 2 global things remain: 'Object' and 'newclass'

-- end of code


lib/HyperKey/utils/on_modifier_hold.lua
local function onModifierHold(modifiers, timeoutMs, onHold, onRelease)
  local state = {
    held = false,
    holdTimer = nil,
    onHold = onHold,
    onRelease = onRelease,
  }

  local cancelTimer = function()
    if not state.holdTimer then return end

    state.holdTimer:stop()
    state.holdTimer = nil
  end

  state.tap = hs.eventtap.new(
    {
      hs.eventtap.event.types.flagsChanged,
    },
    function(event)
      local type = event:getType()
      local containsFlags = event:getFlags():containExactly(modifiers)

      if state.held then
        -- waiting for a release
        if not containsFlags then
          state.held = false
          onRelease()
          cancelTimer()
        end
      elseif state.holdTimer then
        -- we are waiting for the timeout timer to fire
        if not containsFlags then
          cancelTimer()
        end
      elseif containsFlags then
        state.holdTimer = hs.timer.doAfter(timeoutMs / 1000, function()
          state.held = true
          onHold()
        end)
      end

      return false
    end
  )

  state.tap:start()

  return state
end

return onModifierHold


lib/HyperKey/overlay.lua
local newclass = require('HyperKey.utils.yaci')
local rgba = require('HyperKey.utils.rgba')

local Overlay = newclass('Overlay')

local fadeTime = 150 / 1000

function Overlay:init(bindings)
  self.bindings = bindings
  self.canvas = nil
end

function Overlay:show()
  self.canvas = self:_buildCanvas()
  self.canvas:show(fadeTime)
end

function Overlay:hide()
  self.canvas:delete(fadeTime)
  self.canvas = nil
end

function Overlay:_buildCanvas()
  local layerIndexes = {
    background = 1,
    keyIndex = 2,
  }

  table.sort(self.bindings, function(a, b)
    return a.key < b.key
  end)

  -- how much padding around the edges
  local containerPadding = 25

  local itemsPerColumn = 6
  local itemHeight = 25
  local itemBottomMargin = 10
  local itemContainer = itemHeight + itemBottomMargin

  local columnWidth = 275
  local columnCount = math.ceil(#self.bindings / itemsPerColumn)

  local defaultWidth = (containerPadding * 2) + (columnCount * columnWidth)
  local defaultHeight = (containerPadding * 2) + (itemsPerColumn * itemContainer) - itemBottomMargin

  local canvas = hs.canvas.new{
    w = defaultWidth,
    h = defaultHeight,
    x = 100,
    y = 100,
  }

  -- show in center
  local frame = hs.screen.mainScreen():frame()

  canvas:level("overlay")
  canvas:topLeft({
    x = (frame.w / 2) - (defaultWidth / 2),
    y = (frame.h / 2) - (defaultHeight / 2),
  })

  -- render the background rectangle
  canvas:insertElement(
    {
      type = 'rectangle',
      action = 'fill',
      roundedRectRadii = { xRadius = 10, yRadius = 10 },
      fillColor = rgba(24, 135, 250, 1),
      strokeColor = { white = 1.0 },
      strokeWidth = 3.0,
      frame = { x = "0%", y = "0%", h = "100%", w = "100%", },
    },
    layerIndexes.background
  )

  local currentLayerIndex = layerIndexes.keyIndex

  for index, entry in pairs(self.bindings) do
    local zeroIndex = index - 1

    local keySize = 25
    local keyRightMargin = 10

    local columnIndex = math.floor(zeroIndex / itemsPerColumn)

    local startX = containerPadding + (columnIndex * columnWidth)
    local startY = containerPadding + ((zeroIndex % itemsPerColumn) * itemContainer)

    canvas:insertElement(
      {
        type = 'rectangle',
        action = 'fill',
        roundedRectRadii = { xRadius = 5, yRadius = 5 },
        fillColor = rgba(255, 255, 255, 1.0),
        frame = {
          x = startX,
          y = startY,
          w = keySize,
          h = keySize,
        },
        withShadow = true,
        shadow = {
          blurRadius = 5.0,
          color = { alpha = 1/3 },
          offset = { h = -2.0, w = 2.0 },
        }
      },
      currentLayerIndex
    )

    currentLayerIndex = currentLayerIndex + 1

    canvas:insertElement(
      {
        type = 'text',
        text = entry.key,
        action = 'fill',
        frame = {
          x = startX,
          y = startY + 3,
          h = keySize,
          w = keySize,
        },
        textAlignment = "center",
        textColor = rgba(38, 38, 38, 1.0),
        textFont = "Helvetica Bold",
        textSize = 14,
      },
      currentLayerIndex
    )

    currentLayerIndex = currentLayerIndex + 1

    canvas:insertElement(
      {
        type = 'text',
        text = hs.styledtext.new(
          entry.binding.name,
          {
            font = { name = "Helvetica Neue", size = 16 },
            color = rgba(255, 255, 255, 1.0),
            kerning = 1.2,
            shadow = {
              blurRadius = 10,
            }
          }
        ),
        action = 'fill',
        frame = {
          x = startX + keySize + keyRightMargin,
          y = startY,
          h = keySize,
          w = 300,
        },
      },
      currentLayerIndex
    )

    currentLayerIndex = currentLayerIndex + 1
  end

  return canvas
end

return Overlay
